wave_matching/include/wave/matching/pointcloud_display.hpp

```diff
@@ -0,0 +1,121 @@
+/** @file
+ * @ingroup matching
+ *
+ * Interface to PCL Visualiser Class
+ *
+ * This is a class that provides an interface to PCLVisualiser class for
+ * making cool graphics and debugging. The visualiser itself runs in a
+ * separate thread so that it is always interactive.
+ * Note: Ids are shared between all drawn objects, so the should be
+ * globally unique.
+ */
+
+#ifndef WAVE_POINTCLOUDDISPLAY_HPP
+#define WAVE_POINTCLOUDDISPLAY_HPP
+
+#include <atomic>
+#include <queue>
+#include <boost/thread.hpp>
+#include <boost/thread/mutex.hpp>
+#include <pcl/common/common_headers.h>
+#include <pcl/visualization/pcl_visualizer.h>
+#include <pcl/console/parse.h>
+#include "wave/matching/pcl_common.hpp"
+
+namespace wave {
+/** @addtogroup matching
+ *  @{ */
+
+class PointcloudDisplay {
+ public:
+    /**
+     * Default constructor
+     */
+    PointcloudDisplay();
+
+    /**
+     * Calling this launches the worker thread and opens
+     * the visualization window
+     */
+    void startSpin();
+
+    /**
+     * Calling this terminates the worker thread and
+     * should close the window (note there is a bug in
+     * vtk that causes the window to hang around)
+     */
+    void stopSpin();
+
+    /**
+     * Queues a pointcloud to be added to the display.
+     * If worker thread isn't running nothing will happen.
+     * @param cld: cloud to be added
+     * @param id: id of cloud to be added. Same id can be
+     * used to update cloud later
+     */
+    void addPointcloud(PCLPointCloud cld, int id);
+
+    /**
+     * Queues a line to be added to the display.
+     * Note that three objects are added for one line:
+     * spheres at the endpoints, and a line between them.
+     * If worker thread isn't running nothing will happen.
+     * @param pt1 Start coordinate of line
+     * @param pt2 End coordinate of line
+     * @param id1 Start point ID
+     * @param id2 End point ID
+     */
+    void addLine(pcl::PointXYZ pt1, pcl::PointXYZ pt2, int id1, int id2);
+
+ private:
+    /**
+     * PCL Visualizer class instance
+     */
+    boost::shared_ptr<pcl::visualization::PCLVisualizer> viewer;
+    /**
+     * Transfers objects from queues to the internal visualizer class
+     */
+    void updateInternal();
+    /**
+     * Function run by worker class.
+     */
+    void spin();
+    /**
+     * Pointer to worker thread
+     */
+    boost::thread *viewer_thread;
+    /**
+     * Atomic flag. Used by stopSpin to stop worker thread
+     */
+    std::atomic_flag continueFlag = ATOMIC_FLAG_INIT;
+    /**
+     * Mutex to protect buffers
+     */
+    boost::mutex update_mutex;
```

Comment:
Same with above comment, unless you have a good reason why C++'s threads and mutex objects won't do, a comment would be nice. A couple of these doxygen docs are not actually required, when I see the name `mutex` I know its a mutex. or if a variable is a thread I know its a viewer thread. 

If you have to comment a member variable I prefer to see:

    boost::mutex update_mutex;  ///< Mutex to protect buffers